In [1]:
# loading dataset 

import pandas as pd

df = pd.read_csv('/Users/priyamvadakumar/Desktop/compmed/BRCA.exp.547.med.txt', sep ="\t")
print(df)


          NAME  TCGA-A1-A0SD-01A-11R-A115-07  TCGA-A1-A0SE-01A-11R-A084-07  \
0        ELMO2                      0.220167                     -0.105500   
1      CREB3L1                      0.543750                      0.000000   
2        RPS11                      0.354875                      0.305875   
3        PNMA1                      0.433500                      0.039250   
4         MMP2                      1.065167                      0.517000   
...        ...                           ...                           ...   
17809  PIK3IP1                     -0.157500                      0.686000   
17810  SLC39A6                      0.190700                      1.599800   
17811   SNRPD2                      0.047200                     -0.096900   
17812     AQP7                     -0.192500                     -0.546000   
17813     CTSC                     -0.719300                     -0.880500   

       TCGA-A1-A0SH-01A-11R-A084-07  TCGA-A1-A0SJ-01A-11R-A084-

In [2]:
# display columns  
print(df.columns)

Index(['NAME', 'TCGA-A1-A0SD-01A-11R-A115-07', 'TCGA-A1-A0SE-01A-11R-A084-07',
       'TCGA-A1-A0SH-01A-11R-A084-07', 'TCGA-A1-A0SJ-01A-11R-A084-07',
       'TCGA-A1-A0SK-01A-12R-A084-07', 'TCGA-A1-A0SM-01A-11R-A084-07',
       'TCGA-A1-A0SO-01A-22R-A084-07', 'TCGA-A2-A04N-01A-11R-A115-07',
       'TCGA-A2-A04P-01A-31R-A034-07',
       ...
       'TCGA-E2-A15S-01A-11R-A115-07', 'TCGA-E2-A15T-01A-11R-A115-07',
       'TCGA-E2-A1AZ-01A-11R-A12P-07', 'TCGA-E2-A1B0-01A-11R-A12P-07',
       'TCGA-E2-A1B1-01A-21R-A12P-07', 'TCGA-E2-A1B4-01A-11R-A12P-07',
       'TCGA-E2-A1B5-01A-21R-A12P-07', 'TCGA-E2-A1B6-01A-31R-A12P-07',
       'TCGA-E2-A1BC-01A-11R-A12P-07', 'TCGA-E2-A1BD-01A-11R-A12P-07'],
      dtype='object', length=548)


In [3]:
#display samples, called NAME
samples=df.columns[1:]
print(samples)


Index(['TCGA-A1-A0SD-01A-11R-A115-07', 'TCGA-A1-A0SE-01A-11R-A084-07',
       'TCGA-A1-A0SH-01A-11R-A084-07', 'TCGA-A1-A0SJ-01A-11R-A084-07',
       'TCGA-A1-A0SK-01A-12R-A084-07', 'TCGA-A1-A0SM-01A-11R-A084-07',
       'TCGA-A1-A0SO-01A-22R-A084-07', 'TCGA-A2-A04N-01A-11R-A115-07',
       'TCGA-A2-A04P-01A-31R-A034-07', 'TCGA-A2-A04Q-01A-21R-A034-07',
       ...
       'TCGA-E2-A15S-01A-11R-A115-07', 'TCGA-E2-A15T-01A-11R-A115-07',
       'TCGA-E2-A1AZ-01A-11R-A12P-07', 'TCGA-E2-A1B0-01A-11R-A12P-07',
       'TCGA-E2-A1B1-01A-21R-A12P-07', 'TCGA-E2-A1B4-01A-11R-A12P-07',
       'TCGA-E2-A1B5-01A-21R-A12P-07', 'TCGA-E2-A1B6-01A-31R-A12P-07',
       'TCGA-E2-A1BC-01A-11R-A12P-07', 'TCGA-E2-A1BD-01A-11R-A12P-07'],
      dtype='object', length=547)


In [4]:
# extract number representing tumor or non-tumor from the TCGA barcodes representing the samples

for sample in samples:
    code=sample[13:15]
        
    
    print(code)

01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
11
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
11
01
01
01
01
11
0

In [5]:
# check if there are any control samples
print(int(sample[13:15])>19)

False


In [6]:
# create a list of nature ( tumor or non-tumor): code for tumor=1 and code for non-tumor(normal)=0
# ignored the codes for control samples
nature=[]
for sample in samples:
   
    code=int(sample[13:15])
    if code <10:
        target_code=1
    else:
        target_code=0
    nature.append(target_code)
print(nature)

[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 

In [7]:
from collections import Counter


Counter(nature).keys() 
Counter(nature).values() 

dict_values([525, 22])

In [8]:
# change the index 0, 1, 2 etc to the NAME values before doing the transpose
# for making features in the columns and samples in the rows

df.set_index('NAME', inplace=True)
print(df)

         TCGA-A1-A0SD-01A-11R-A115-07  TCGA-A1-A0SE-01A-11R-A084-07  \
NAME                                                                  
ELMO2                        0.220167                     -0.105500   
CREB3L1                      0.543750                      0.000000   
RPS11                        0.354875                      0.305875   
PNMA1                        0.433500                      0.039250   
MMP2                         1.065167                      0.517000   
...                               ...                           ...   
PIK3IP1                     -0.157500                      0.686000   
SLC39A6                      0.190700                      1.599800   
SNRPD2                       0.047200                     -0.096900   
AQP7                        -0.192500                     -0.546000   
CTSC                        -0.719300                     -0.880500   

         TCGA-A1-A0SH-01A-11R-A084-07  TCGA-A1-A0SJ-01A-11R-A084-07  \
NAME 

In [9]:
#now doing the transpose

data=df.T
print(data)
data.shape

NAME                             ELMO2  CREB3L1     RPS11    PNMA1      MMP2  \
TCGA-A1-A0SD-01A-11R-A115-07  0.220167  0.54375  0.354875  0.43350  1.065167   
TCGA-A1-A0SE-01A-11R-A084-07 -0.105500  0.00000  0.305875  0.03925  0.517000   
TCGA-A1-A0SH-01A-11R-A084-07  0.174583  1.36850  0.007000  0.23250  1.844667   
TCGA-A1-A0SJ-01A-11R-A084-07  0.590083 -0.45300 -0.295125  0.66525 -0.530167   
TCGA-A1-A0SK-01A-12R-A084-07  1.125417 -1.52800  0.082750  0.85075 -1.674000   
...                                ...      ...       ...      ...       ...   
TCGA-E2-A1B4-01A-11R-A12P-07 -0.638083  0.82750 -0.084000 -0.45600 -0.040167   
TCGA-E2-A1B5-01A-21R-A12P-07 -0.497333 -0.96250  0.193500 -1.33700 -1.446833   
TCGA-E2-A1B6-01A-31R-A12P-07 -0.324000  0.50950  0.996500 -0.51375 -0.658000   
TCGA-E2-A1BC-01A-11R-A12P-07 -0.188083  0.73950  0.239625 -0.54225  1.196333   
TCGA-E2-A1BD-01A-11R-A12P-07 -0.601917 -0.42550  0.190625 -0.29475  1.064167   

NAME                          C10orf90 

(547, 17814)

In [10]:
# create response varibale from the nature which is list representing tumor or otherwise in the form code
data['response']=nature

print(data['response'])

TCGA-A1-A0SD-01A-11R-A115-07    1
TCGA-A1-A0SE-01A-11R-A084-07    1
TCGA-A1-A0SH-01A-11R-A084-07    1
TCGA-A1-A0SJ-01A-11R-A084-07    1
TCGA-A1-A0SK-01A-12R-A084-07    1
                               ..
TCGA-E2-A1B4-01A-11R-A12P-07    1
TCGA-E2-A1B5-01A-21R-A12P-07    1
TCGA-E2-A1B6-01A-31R-A12P-07    1
TCGA-E2-A1BC-01A-11R-A12P-07    1
TCGA-E2-A1BD-01A-11R-A12P-07    1
Name: response, Length: 547, dtype: int64


In [11]:
print(data['response'])

TCGA-A1-A0SD-01A-11R-A115-07    1
TCGA-A1-A0SE-01A-11R-A084-07    1
TCGA-A1-A0SH-01A-11R-A084-07    1
TCGA-A1-A0SJ-01A-11R-A084-07    1
TCGA-A1-A0SK-01A-12R-A084-07    1
                               ..
TCGA-E2-A1B4-01A-11R-A12P-07    1
TCGA-E2-A1B5-01A-21R-A12P-07    1
TCGA-E2-A1B6-01A-31R-A12P-07    1
TCGA-E2-A1BC-01A-11R-A12P-07    1
TCGA-E2-A1BD-01A-11R-A12P-07    1
Name: response, Length: 547, dtype: int64


In [12]:
print(data)

NAME                             ELMO2  CREB3L1     RPS11    PNMA1      MMP2  \
TCGA-A1-A0SD-01A-11R-A115-07  0.220167  0.54375  0.354875  0.43350  1.065167   
TCGA-A1-A0SE-01A-11R-A084-07 -0.105500  0.00000  0.305875  0.03925  0.517000   
TCGA-A1-A0SH-01A-11R-A084-07  0.174583  1.36850  0.007000  0.23250  1.844667   
TCGA-A1-A0SJ-01A-11R-A084-07  0.590083 -0.45300 -0.295125  0.66525 -0.530167   
TCGA-A1-A0SK-01A-12R-A084-07  1.125417 -1.52800  0.082750  0.85075 -1.674000   
...                                ...      ...       ...      ...       ...   
TCGA-E2-A1B4-01A-11R-A12P-07 -0.638083  0.82750 -0.084000 -0.45600 -0.040167   
TCGA-E2-A1B5-01A-21R-A12P-07 -0.497333 -0.96250  0.193500 -1.33700 -1.446833   
TCGA-E2-A1B6-01A-31R-A12P-07 -0.324000  0.50950  0.996500 -0.51375 -0.658000   
TCGA-E2-A1BC-01A-11R-A12P-07 -0.188083  0.73950  0.239625 -0.54225  1.196333   
TCGA-E2-A1BD-01A-11R-A12P-07 -0.601917 -0.42550  0.190625 -0.29475  1.064167   

NAME                          C10orf90 

In [13]:
print(data.head())

NAME                             ELMO2  CREB3L1     RPS11    PNMA1      MMP2  \
TCGA-A1-A0SD-01A-11R-A115-07  0.220167  0.54375  0.354875  0.43350  1.065167   
TCGA-A1-A0SE-01A-11R-A084-07 -0.105500  0.00000  0.305875  0.03925  0.517000   
TCGA-A1-A0SH-01A-11R-A084-07  0.174583  1.36850  0.007000  0.23250  1.844667   
TCGA-A1-A0SJ-01A-11R-A084-07  0.590083 -0.45300 -0.295125  0.66525 -0.530167   
TCGA-A1-A0SK-01A-12R-A084-07  1.125417 -1.52800  0.082750  0.85075 -1.674000   

NAME                          C10orf90      ZHX3    ERCC5     GPR98   RXFP3  \
TCGA-A1-A0SD-01A-11R-A115-07  -0.11225  0.262333 -0.16625  0.322375 -0.4960   
TCGA-A1-A0SE-01A-11R-A084-07   1.31575  0.175333 -0.29750  0.067000  0.0085   
TCGA-A1-A0SH-01A-11R-A084-07   0.17925  0.639500 -0.05175 -0.476750 -0.1380   
TCGA-A1-A0SJ-01A-11R-A084-07   0.12225  0.470000 -0.02375 -0.421375  0.2545   
TCGA-A1-A0SK-01A-12R-A084-07  -0.84900 -1.112500  0.73200  0.233750 -0.4940   

NAME                          ...    GRIP2  

In [14]:
data.describe()

NAME        ELMO2     CREB3L1       RPS11       PNMA1        MMP2    C10orf90  \
count  547.000000  547.000000  547.000000  547.000000  547.000000  547.000000   
mean    -0.000167   -0.005408    0.018381   -0.065037   -0.074444    0.252379   
std      0.547277    0.819250    0.503550    0.609022    0.957383    0.980748   
min     -1.811833   -2.303500   -1.436000   -1.818750   -2.725500   -3.120750   
25%     -0.366125   -0.621000   -0.320438   -0.464500   -0.752417   -0.315125   
50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
75%      0.313750    0.567125    0.312250    0.356000    0.635000    0.434416   
max      2.614250    2.329500    2.111875    1.508000    2.360333    4.478500   

NAME         ZHX3       ERCC5       GPR98       RXFP3  ...       GRIP2  \
count  547.000000  547.000000  547.000000  547.000000  ...  547.000000   
mean     0.007024   -0.000391   -0.166210    0.008223  ...   -0.017261   
std      0.469728    0.490774    1.132159    0.367285  ...    0.330794   
min     -1.518500   -1.570250   -2.587750   -1.634000  ...   -1.803250   
25%     -0.255167   -0.327250   -1.179937   -0.204500  ...   -0.171125   
50%      0.000000    0.000000    0.000000    0.000000  ...    0.000000   
75%      0.297167    0.349375    0.706500    0.189500  ...    0.173250   
max      1.580667    1.882750    2.600679    2.119000  ...    1.468250   

NAME        GPLD1       RAB8A       RXFP2     PIK3IP1     SLC39A6      SNRPD2  \
count  547.000000  547.000000  546.000000  547.000000  547.000000  547.000000   
mean     0.098146    0.004173    0.012682   -0.083575    0.162335    0.016307   
std      0.821084    0.518305    0.681432    0.676153    1.794338    0.423155   
min     -1.995667   -1.675500   -3.167500   -2.873000   -4.187600   -1.105100   
25%     -0.471583   -0.330150   -0.327000   -0.449875   -1.267950   -0.274750   
50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
75%      0.602833    0.334850    0.373875    0.393500    1.549500    0.251200   
max      3.482833    2.047800    4.365000    1.864500    4.282100    1.770100   

NAME         AQP7        CTSC    response  
count  547.000000  547.000000  547.000000  
mean     0.367138    0.055690    0.959781  
std      1.220861    0.852395    0.196653  
min     -2.064500   -1.876800    0.000000  
25%     -0.417000   -0.539950    1.000000  
50%      0.000000    0.000000    1.000000  
75%      0.806250    0.582750    1.000000  
max      5.343500    3.114400    1.000000  

[8 rows x 17815 columns]

In [15]:
# check if any NaN values 

In [16]:
data.isnull().values.any()

True

In [17]:
data.isnull().sum().sum()

1570

In [18]:
# columns that have null values
data.columns[data.isna().any()].tolist()

['OR2K2',
 'ADAM5P',
 'KIAA1486',
 'C9orf11',
 'C7orf33',
 'OR4C13',
 'DEFB125',
 'SEBOX',
 'CMA1',
 'CCDC60',
 'KRT74',
 'MAS1',
 'LCE1E',
 'OR8B4',
 'NEU2',
 'FLJ32894',
 'EGFL11',
 'CXorf59',
 'PTH',
 'SOX6',
 'NMUR2',
 'LIPK',
 'TBX22',
 'UNQ6190',
 'IGFL1',
 'RTP3',
 'SPO11',
 'GABRR2',
 'GPR22',
 'ACTRT1',
 'KIR3DL2',
 'OR1G1',
 'PAGE2B',
 'CDH17',
 'PCDHA9',
 'RNASE9',
 'ACCN5',
 'PPP1R3A',
 'MRGPRX1',
 'OR4A5',
 'LIPF',
 'LHCGR',
 'APOBEC1',
 'MT4',
 'FBXL21',
 'DEFB127',
 'C6orf195',
 'ATP13A5',
 'C10orf111',
 'FRMD7',
 'GPR111',
 'SERPINB12',
 'CPLX4',
 'KRT26',
 'AADACL4',
 'NLRP10',
 'NAG18',
 'MMP20',
 'FLJ40292',
 'EDAR',
 'CNTN5',
 'OR10A6',
 'HTR3D',
 'UNQ9391',
 'SPRYD5',
 'PCDHGB7',
 'WBP2NL',
 'HOXD12',
 'TAS2R42',
 'DEFB118',
 'C18orf26',
 'LRIT2',
 'TAS2R8',
 'C15orf32',
 'TRIML1',
 'TSGA13',
 'KNCN',
 'NXPH2',
 'KRT75',
 'ARSF',
 'GCM2',
 'S100G',
 'KIR3DX1',
 'OR11H6',
 'PRKAG3',
 'KY',
 'FAM55D',
 'C4orf40',
 'C6orf146',
 'GSX2',
 'ZNF674',
 'AQP12B',
 'C16orf46

In [19]:
# imputing for all NaN values -> conversion to 0
data = data.fillna(0)




In [20]:
data.isnull().values.any()

False

In [21]:
# now all NaNs have been replaced by 0 

In [22]:
features=data.iloc[:,:-1]
print(features)
print(features.shape)

NAME                             ELMO2  CREB3L1     RPS11    PNMA1      MMP2  \
TCGA-A1-A0SD-01A-11R-A115-07  0.220167  0.54375  0.354875  0.43350  1.065167   
TCGA-A1-A0SE-01A-11R-A084-07 -0.105500  0.00000  0.305875  0.03925  0.517000   
TCGA-A1-A0SH-01A-11R-A084-07  0.174583  1.36850  0.007000  0.23250  1.844667   
TCGA-A1-A0SJ-01A-11R-A084-07  0.590083 -0.45300 -0.295125  0.66525 -0.530167   
TCGA-A1-A0SK-01A-12R-A084-07  1.125417 -1.52800  0.082750  0.85075 -1.674000   
...                                ...      ...       ...      ...       ...   
TCGA-E2-A1B4-01A-11R-A12P-07 -0.638083  0.82750 -0.084000 -0.45600 -0.040167   
TCGA-E2-A1B5-01A-21R-A12P-07 -0.497333 -0.96250  0.193500 -1.33700 -1.446833   
TCGA-E2-A1B6-01A-31R-A12P-07 -0.324000  0.50950  0.996500 -0.51375 -0.658000   
TCGA-E2-A1BC-01A-11R-A12P-07 -0.188083  0.73950  0.239625 -0.54225  1.196333   
TCGA-E2-A1BD-01A-11R-A12P-07 -0.601917 -0.42550  0.190625 -0.29475  1.064167   

NAME                          C10orf90 

In [23]:
print(type(features))

<class 'pandas.core.frame.DataFrame'>


In [24]:
# convert features into numpy to enable use of sklearn
X=features.to_numpy()
print(type(X))
print(X)

<class 'numpy.ndarray'>
[[ 0.220167  0.54375   0.354875 ...  0.0472   -0.1925   -0.7193  ]
 [-0.1055    0.        0.305875 ... -0.0969   -0.546    -0.8805  ]
 [ 0.174583  1.3685    0.007    ... -0.2257   -0.613    -0.491   ]
 ...
 [-0.324     0.5095    0.9965   ...  0.2874    2.1415    0.9334  ]
 [-0.188083  0.7395    0.239625 ... -0.5966    3.9995   -0.8675  ]
 [-0.601917 -0.4255    0.190625 ...  0.1374    0.1525   -0.859   ]]


In [25]:
print(X.shape) # 547 samples and 17814 genes 

(547, 17814)


In [26]:

y=data['response'].to_numpy()


In [27]:
print(y)

[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 0
 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 0 1 1 1 1 0 1 0 1 1 0 1 1 1 1 1 0 1
 1 1 0 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 0 1 0 1 0 1 1 1 0 1 1 0 1 1 0 1 1
 1 1 0 1 0 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 

In [28]:
#train, test split

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=1, stratify=y)
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(366, 17814) (181, 17814) (366,) (181,)


In [29]:
print(y_train)

[1 1 1 1 1 1 1 0 1 1 1 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]


In [30]:
y_test

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 0])

In [31]:
# normalizing  X_train and X_test, separately 

from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
features_train_scaled = pd.DataFrame(scaler.fit_transform(X_train),columns = features.columns)
features_test_scaled = pd.DataFrame(scaler.fit_transform(X_test),columns = features.columns)


In [32]:
print(type(features_train_scaled))
print(features_train_scaled.shape)
print(features_train_scaled)

<class 'pandas.core.frame.DataFrame'>
(366, 17814)
NAME     ELMO2   CREB3L1     RPS11     PNMA1      MMP2  C10orf90      ZHX3  \
0     0.352701  0.187460  0.246745  0.600136  0.243443  0.323228  0.470826   
1     0.289988  0.720591  0.605208  0.411259  0.577570  0.369501  0.489002   
2     0.414947  0.572847  0.468159  0.499623  0.618318  0.628213  0.460984   
3     0.400144  0.546784  0.361975  0.632131  0.817392  0.303233  0.434418   
4     0.278103  0.390622  0.197467  0.068292  0.405627  0.280116  0.506857   
..         ...       ...       ...       ...       ...       ...       ...   
361   0.590714  0.336391  0.359811  0.526336  0.548763  0.324561  0.249906   
362   0.214768  0.564537  0.377302  0.410504  0.698605  0.399779  0.494219   
363   0.452508  0.598262  0.461951  0.902883  0.465357  0.243783  0.563323   
364   0.329359  0.719242  0.629049  0.442650  0.668053  0.382641  0.423178   
365   0.338131  0.429473  0.439884  0.740794  0.675029  0.380051  0.343318   

NAME     ERC

In [33]:
features_test_scaled

NAME     ELMO2   CREB3L1     RPS11     PNMA1      MMP2  C10orf90      ZHX3  \
0     0.486079  0.367803  0.213673  0.500763  0.424482  0.376186  0.457033   
1     0.225141  0.617157  0.588138  0.554141  0.534980  0.368519  0.761242   
2     0.400642  0.582236  0.308684  0.715876  0.707446  0.381929  0.343104   
3     0.397787  0.377227  0.474852  0.185298  0.052379  1.000000  0.673368   
4     0.378963  0.485347  0.270801  0.514107  0.555685  0.424388  0.673619   
..         ...       ...       ...       ...       ...       ...       ...   
176   0.436912  0.694810  0.411822  0.397057  0.741700  0.456578  0.386634   
177   0.424172  0.474890  0.107100  0.538661  0.561460  0.430569  0.501002   
178   0.429370  0.572747  0.291542  0.759036  0.784853  0.360311  0.461042   
179   0.397669  0.807255  0.360314  0.399344  1.000000  0.524607  0.318427   
180   0.000000  0.245223  0.679510  0.589751  0.252452  0.640196  0.491795   

NAME     ERCC5     GPR98     RXFP3  ...   SLC2A11     GRIP2     GPLD1  \
0     0.441083  0.390708  0.390221  ...  0.492849  0.567690  0.633867   
1     0.710911  0.628541  0.554756  ...  0.644484  0.404401  0.020048   
2     0.744617  0.564738  0.430189  ...  0.595656  0.509207  0.272361   
3     0.577996  0.367586  0.621370  ...  0.611114  0.524030  0.464231   
4     0.511765  0.563956  0.374634  ...  0.631820  0.588882  0.309673   
..         ...       ...       ...  ...       ...       ...       ...   
176   0.566367  0.537164  0.574607  ...  0.450041  0.400232  0.151816   
177   0.495685  0.642035  0.467626  ...  0.518659  0.631847  0.532642   
178   0.599164  0.694177  0.371703  ...  0.517070  0.458946  0.175934   
179   0.647043  0.026694  0.519851  ...  0.473588  0.524493  0.267949   
180   0.661397  0.452607  0.306422  ...  0.552704  0.574870  0.745288   

NAME     RAB8A     RXFP2   PIK3IP1   SLC39A6    SNRPD2      AQP7      CTSC  
0     0.480591  0.218426  0.593053  0.716536  0.552622  0.000000  0.450890  
1     0.458235  0.440064  0.677306  0.578588  0.524242  0.022071  0.285755  
2     0.390425  0.407140  0.489878  0.581102  0.303805  0.153995  0.144091  
3     0.501296  0.155716  0.609501  0.130200  0.706838  0.357363  0.600202  
4     0.571669  0.251643  0.436738  0.701209  0.701864  0.226729  0.217260  
..         ...       ...       ...       ...       ...       ...       ...  
176   0.607524  0.508614  0.733379  0.541014  0.434126  0.105912  0.337313  
177   0.725486  0.417214  0.484357  0.835050  0.374965  0.140523  0.378389  
178   0.390852  0.433348  0.577525  0.391055  0.290623  0.133429  0.477459  
179   0.538689  0.501095  0.726478  0.332621  0.221863  0.646220  0.302206  
180   0.000000  0.342386  0.501668  0.405385  0.569978  1.000000  0.443039  

[181 rows x 17814 columns]

In [34]:
features_train_scaled[:2]

NAME     ELMO2   CREB3L1     RPS11     PNMA1      MMP2  C10orf90      ZHX3  \
0     0.352701  0.187460  0.246745  0.600136  0.243443  0.323228  0.470826   
1     0.289988  0.720591  0.605208  0.411259  0.577570  0.369501  0.489002   

NAME     ERCC5     GPR98     RXFP3  ...   SLC2A11     GRIP2     GPLD1  \
0     0.423827  0.664747  0.365564  ...  0.561410  0.472566  0.349396   
1     0.463700  0.692406  0.380545  ...  0.289275  0.511233  0.307809   

NAME     RAB8A     RXFP2   PIK3IP1   SLC39A6    SNRPD2      AQP7      CTSC  
0     0.387747  0.485108  0.716095  0.819521  0.240309  0.241881  0.229866  
1     0.325578  0.302615  0.769436  0.439354  0.574545  0.587523  0.611235  

[2 rows x 17814 columns]

In [35]:
# # checking column number of a random gene "ZHX3" to make sure data is in order

# index_no = features_train_scaled.columns.get_loc('ZHX3')
# print(index_no)

## Feature Filtering here - ANOVA F-test 

In [57]:
#ANOVA based filtering


from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif
from matplotlib import pyplot

# feature selection
def select_features(X_train, y_train):
    # configure to select all features
    fs = SelectKBest(score_func=f_classif, k='all')
    # learn relationship from training data
    fs.fit(X_train, y_train)
    # transform train input data
    X_train_fs = fs.transform(X_train)
    # transform test input data
    #X_test_fs = fs.transform(X_test)
    return X_train_fs, fs


# feature selection
X_train_fs, fs = select_features(features_train_scaled, y_train)
# what are scores for the features

for i in range(len(fs.scores_)):
    print('Feature %d: %f' % (i, fs.scores_[i]))


        


Feature 0: 12.975122
Feature 1: 28.172079
Feature 2: 24.621675
Feature 3: 1.841134
Feature 4: 0.775998
Feature 5: 74.308357
Feature 6: 0.500364
Feature 7: 10.213869
Feature 8: 0.037326
Feature 9: 5.798746
Feature 10: 1.046235
Feature 11: 11.383435
Feature 12: 75.275412
Feature 13: 1.170789
Feature 14: 5.653846
Feature 15: 1.073641
Feature 16: 0.223971
Feature 17: 12.579034
Feature 18: 14.933094
Feature 19: 6.159668
Feature 20: 0.353899
Feature 21: 0.243082
Feature 22: 0.296687
Feature 23: 13.903142
Feature 24: 1.497335
Feature 25: 3.998311
Feature 26: 6.252610
Feature 27: 0.009050
Feature 28: 15.161697
Feature 29: 28.139123
Feature 30: 7.853151
Feature 31: 10.993556
Feature 32: 5.685363
Feature 33: 0.074205
Feature 34: 4.285267
Feature 35: 18.807216
Feature 36: 4.962639
Feature 37: 0.554521
Feature 38: 4.709638
Feature 39: 1.745883
Feature 40: 3.874338
Feature 41: 1.995823
Feature 42: 0.058396
Feature 43: 0.320390
Feature 44: 0.582367
Feature 45: 0.399824
Feature 46: 15.234742
Feature 

Feature 2186: 1.281902
Feature 2187: 14.638105
Feature 2188: 7.511856
Feature 2189: 20.141618
Feature 2190: 1.067794
Feature 2191: 8.105691
Feature 2192: 1.149910
Feature 2193: 2.861051
Feature 2194: 0.441610
Feature 2195: 0.462148
Feature 2196: 2.123542
Feature 2197: 0.024180
Feature 2198: 0.376163
Feature 2199: 0.177484
Feature 2200: 0.507952
Feature 2201: 0.292485
Feature 2202: 7.757189
Feature 2203: 3.968018
Feature 2204: 12.333342
Feature 2205: 7.429931
Feature 2206: 14.457280
Feature 2207: 0.016763
Feature 2208: 4.043345
Feature 2209: 16.022892
Feature 2210: 26.697530
Feature 2211: 0.870265
Feature 2212: 0.599149
Feature 2213: 16.353548
Feature 2214: 5.426604
Feature 2215: 0.459126
Feature 2216: 0.536879
Feature 2217: 0.002272
Feature 2218: 10.202851
Feature 2219: 3.666078
Feature 2220: 2.245244
Feature 2221: 1.428481
Feature 2222: 1.369619
Feature 2223: 8.715067
Feature 2224: 10.258896
Feature 2225: 13.951885
Feature 2226: 6.387783
Feature 2227: 1.682520
Feature 2228: 1.877687
F

Feature 3220: 0.116479
Feature 3221: 178.564264
Feature 3222: 1.642791
Feature 3223: 20.319889
Feature 3224: 0.056571
Feature 3225: 1.425266
Feature 3226: 5.752099
Feature 3227: 30.853797
Feature 3228: 22.881398
Feature 3229: 1.190798
Feature 3230: 0.181132
Feature 3231: 1.982782
Feature 3232: 1.131294
Feature 3233: 15.778965
Feature 3234: 5.588660
Feature 3235: 10.756018
Feature 3236: 20.718884
Feature 3237: 31.348973
Feature 3238: 82.846282
Feature 3239: 12.488983
Feature 3240: 8.466335
Feature 3241: 0.444767
Feature 3242: 4.906404
Feature 3243: 27.973591
Feature 3244: 0.142897
Feature 3245: 7.147301
Feature 3246: 0.646940
Feature 3247: 19.858274
Feature 3248: 10.976680
Feature 3249: 1.287998
Feature 3250: 0.611147
Feature 3251: 0.014066
Feature 3252: 0.070425
Feature 3253: 34.297008
Feature 3254: 1.716549
Feature 3255: 0.083227
Feature 3256: 2.930565
Feature 3257: 108.720340
Feature 3258: 164.793987
Feature 3259: 0.177689
Feature 3260: 0.582471
Feature 3261: 0.826195
Feature 3262: 2

Feature 4719: 0.815811
Feature 4720: 0.176696
Feature 4721: 0.320042
Feature 4722: 0.010486
Feature 4723: 5.769134
Feature 4724: 23.962975
Feature 4725: 11.147813
Feature 4726: 102.495149
Feature 4727: 37.605487
Feature 4728: 2.130087
Feature 4729: 26.475181
Feature 4730: 14.100634
Feature 4731: 1.746342
Feature 4732: 0.375552
Feature 4733: 13.746093
Feature 4734: 26.052120
Feature 4735: 1.516538
Feature 4736: 5.626062
Feature 4737: 0.043694
Feature 4738: 12.141280
Feature 4739: 0.593323
Feature 4740: 0.273739
Feature 4741: 15.966287
Feature 4742: 12.685476
Feature 4743: 1.157153
Feature 4744: 0.374123
Feature 4745: 7.431635
Feature 4746: 16.771858
Feature 4747: 1.290960
Feature 4748: 7.761168
Feature 4749: 27.259050
Feature 4750: 1.259832
Feature 4751: 3.793447
Feature 4752: 0.421721
Feature 4753: 25.780258
Feature 4754: 26.034458
Feature 4755: 6.752861
Feature 4756: 0.099739
Feature 4757: 47.570830
Feature 4758: 3.284292
Feature 4759: 20.130696
Feature 4760: 0.138588
Feature 4761: 21

Feature 6719: 20.687277
Feature 6720: 0.230710
Feature 6721: 4.228708
Feature 6722: 13.974196
Feature 6723: 0.236194
Feature 6724: 0.301354
Feature 6725: 12.388762
Feature 6726: 13.996583
Feature 6727: 0.701956
Feature 6728: 52.598112
Feature 6729: 1.906604
Feature 6730: 19.094964
Feature 6731: 2.732435
Feature 6732: 1.824568
Feature 6733: 3.309768
Feature 6734: 19.921425
Feature 6735: 10.031377
Feature 6736: 2.121778
Feature 6737: 7.131257
Feature 6738: 71.015142
Feature 6739: 9.511162
Feature 6740: 50.705885
Feature 6741: 0.877195
Feature 6742: 16.338528
Feature 6743: 1.999378
Feature 6744: 5.906141
Feature 6745: 23.553300
Feature 6746: 23.994216
Feature 6747: 3.783698
Feature 6748: 2.774072
Feature 6749: 8.031629
Feature 6750: 1.153976
Feature 6751: 0.382514
Feature 6752: 10.091127
Feature 6753: 0.812955
Feature 6754: 1.441917
Feature 6755: 3.854979
Feature 6756: 7.302797
Feature 6757: 1.940529
Feature 6758: 17.512679
Feature 6759: 0.009174
Feature 6760: 8.884898
Feature 6761: 0.143

Feature 8718: 7.817605
Feature 8719: 62.954910
Feature 8720: 0.202247
Feature 8721: 1.662445
Feature 8722: 72.050756
Feature 8723: 14.869711
Feature 8724: 5.504370
Feature 8725: 8.295689
Feature 8726: 19.832443
Feature 8727: 38.453324
Feature 8728: 22.084245
Feature 8729: 10.263865
Feature 8730: 11.218786
Feature 8731: 0.868288
Feature 8732: 0.873007
Feature 8733: 3.274000
Feature 8734: 2.889758
Feature 8735: 37.148740
Feature 8736: 3.140312
Feature 8737: 8.129534
Feature 8738: 14.741663
Feature 8739: 9.640319
Feature 8740: 10.742607
Feature 8741: 0.099561
Feature 8742: 1.296466
Feature 8743: 8.632435
Feature 8744: 39.260097
Feature 8745: 8.138073
Feature 8746: 1.200651
Feature 8747: 3.182814
Feature 8748: 28.347814
Feature 8749: 13.473679
Feature 8750: 31.087157
Feature 8751: 4.238156
Feature 8752: 51.337117
Feature 8753: 8.453055
Feature 8754: 68.367696
Feature 8755: 2.196266
Feature 8756: 6.769232
Feature 8757: 0.042067
Feature 8758: 3.644295
Feature 8759: 3.843749
Feature 8760: 28.

Feature 10218: 3.216215
Feature 10219: 34.220793
Feature 10220: 16.015891
Feature 10221: 1.898901
Feature 10222: 20.411051
Feature 10223: 0.592349
Feature 10224: 2.719070
Feature 10225: 0.665124
Feature 10226: 3.603148
Feature 10227: 0.067826
Feature 10228: 0.190304
Feature 10229: 12.754575
Feature 10230: 1.328720
Feature 10231: 2.232526
Feature 10232: 2.504836
Feature 10233: 70.920542
Feature 10234: 0.240769
Feature 10235: 0.023779
Feature 10236: 72.477474
Feature 10237: 29.009874
Feature 10238: 5.494435
Feature 10239: 0.103093
Feature 10240: 3.593647
Feature 10241: 9.169385
Feature 10242: 2.593177
Feature 10243: 54.723920
Feature 10244: 58.201321
Feature 10245: 0.000363
Feature 10246: 0.003570
Feature 10247: 0.627841
Feature 10248: 19.801083
Feature 10249: 8.695832
Feature 10250: 0.047610
Feature 10251: 0.378873
Feature 10252: 2.154818
Feature 10253: 8.435990
Feature 10254: 2.982934
Feature 10255: 0.013338
Feature 10256: 12.540861
Feature 10257: 5.841992
Feature 10258: 2.412493
Featu

Feature 12217: 0.818612
Feature 12218: 0.856508
Feature 12219: 12.953913
Feature 12220: 1.799082
Feature 12221: 0.790136
Feature 12222: 6.048021
Feature 12223: 18.474930
Feature 12224: 0.303786
Feature 12225: 21.728970
Feature 12226: 16.375379
Feature 12227: 5.519528
Feature 12228: 1.543485
Feature 12229: 7.228762
Feature 12230: 0.689937
Feature 12231: 10.374705
Feature 12232: 14.399764
Feature 12233: 21.981629
Feature 12234: 0.183908
Feature 12235: 2.693369
Feature 12236: 5.610019
Feature 12237: 43.088632
Feature 12238: 8.012402
Feature 12239: 0.004223
Feature 12240: 7.641928
Feature 12241: 78.235602
Feature 12242: 9.463991
Feature 12243: 0.549492
Feature 12244: 3.614337
Feature 12245: 12.072601
Feature 12246: 11.756197
Feature 12247: 22.608963
Feature 12248: 15.065807
Feature 12249: 0.681262
Feature 12250: 1.232044
Feature 12251: 18.365882
Feature 12252: 9.482400
Feature 12253: 3.553920
Feature 12254: 0.370498
Feature 12255: 0.015227
Feature 12256: 0.641676
Feature 12257: 21.380375
F

Feature 14018: 9.825503
Feature 14019: 0.242254
Feature 14020: 0.021845
Feature 14021: 2.082539
Feature 14022: 19.776342
Feature 14023: 0.022832
Feature 14024: 0.087122
Feature 14025: 1.015717
Feature 14026: 11.782438
Feature 14027: 3.912620
Feature 14028: 1.958209
Feature 14029: 0.227689
Feature 14030: 2.058638
Feature 14031: 1.079751
Feature 14032: 36.013341
Feature 14033: 6.015423
Feature 14034: 2.300863
Feature 14035: 55.999432
Feature 14036: 16.380213
Feature 14037: 10.344793
Feature 14038: 0.170045
Feature 14039: 1.175159
Feature 14040: 20.296998
Feature 14041: 0.017096
Feature 14042: 0.659169
Feature 14043: 0.033262
Feature 14044: 2.742317
Feature 14045: 0.535534
Feature 14046: 0.009911
Feature 14047: 19.312074
Feature 14048: 5.852194
Feature 14049: 44.355212
Feature 14050: 6.663824
Feature 14051: 3.907060
Feature 14052: 1.001054
Feature 14053: 16.894995
Feature 14054: 2.214402
Feature 14055: 6.657125
Feature 14056: 3.435714
Feature 14057: 10.528106
Feature 14058: 70.211185
Feat

Feature 15716: 23.178483
Feature 15717: 7.517409
Feature 15718: 0.000074
Feature 15719: 19.885574
Feature 15720: 1.375739
Feature 15721: 8.767953
Feature 15722: 0.371782
Feature 15723: 2.548973
Feature 15724: 3.042100
Feature 15725: 6.623061
Feature 15726: 8.319285
Feature 15727: 16.574238
Feature 15728: 3.074245
Feature 15729: 9.496600
Feature 15730: 4.050077
Feature 15731: 4.003963
Feature 15732: 0.220768
Feature 15733: 10.462417
Feature 15734: 25.194722
Feature 15735: 13.844427
Feature 15736: 91.628067
Feature 15737: 20.577929
Feature 15738: 2.214124
Feature 15739: 6.211562
Feature 15740: 4.414383
Feature 15741: 0.282702
Feature 15742: 0.231028
Feature 15743: 10.765382
Feature 15744: 5.186672
Feature 15745: 6.139176
Feature 15746: 5.346617
Feature 15747: 1.504215
Feature 15748: 6.441338
Feature 15749: 0.691756
Feature 15750: 14.571324
Feature 15751: 19.543542
Feature 15752: 44.989305
Feature 15753: 8.297107
Feature 15754: 0.352640
Feature 15755: 10.198837
Feature 15756: 3.231568
Fea

Feature 17716: 0.496267
Feature 17717: 3.691588
Feature 17718: 0.086922
Feature 17719: 0.733647
Feature 17720: 50.717611
Feature 17721: 18.277823
Feature 17722: 1.556942
Feature 17723: 7.780316
Feature 17724: 5.538973
Feature 17725: 29.656728
Feature 17726: 9.823811
Feature 17727: 0.054203
Feature 17728: 0.161016
Feature 17729: 24.344642
Feature 17730: 1.233630
Feature 17731: 14.407966
Feature 17732: 9.387179
Feature 17733: 0.342979
Feature 17734: 0.001154
Feature 17735: 0.045026
Feature 17736: 14.139553
Feature 17737: 2.409798
Feature 17738: 0.007125
Feature 17739: 0.687262
Feature 17740: 10.165229
Feature 17741: 0.950098
Feature 17742: 5.581074
Feature 17743: 2.131901
Feature 17744: 29.371347
Feature 17745: 2.576035
Feature 17746: 0.162750
Feature 17747: 15.837541
Feature 17748: 0.165494
Feature 17749: 8.741585
Feature 17750: 5.258439
Feature 17751: 1.633561
Feature 17752: 0.040120
Feature 17753: 2.157538
Feature 17754: 30.706152
Feature 17755: 3.495585
Feature 17756: 8.296202
Featur

In [58]:
print(X_train_fs.shape)
print(X_train_fs)

(366, 17814)
[[0.35270092 0.18745953 0.24674495 ... 0.24030942 0.24188052 0.22986648]
 [0.28998808 0.72059141 0.60520807 ... 0.57454453 0.58752347 0.61123536]
 [0.41494668 0.57284697 0.46815908 ... 0.2756478  0.66318937 0.53882775]
 ...
 [0.45250827 0.59826246 0.46195054 ... 0.36412056 0.11544614 0.33099356]
 [0.3293587  0.71924239 0.62904885 ... 0.55776303 0.5146394  0.29862049]
 [0.3381314  0.42947334 0.43988363 ... 0.3252737  0.46922595 0.36728593]]


In [59]:
l=[(i, fs.scores_[i]) for i in range(len(fs.scores_))]
print(l) 

[(0, 12.975121587382011), (1, 28.172079199056192), (2, 24.62167518253117), (3, 1.8411343235581346), (4, 0.7759983819998735), (5, 74.30835672354564), (6, 0.5003636681523963), (7, 10.213869447511394), (8, 0.03732592508648201), (9, 5.798746076427377), (10, 1.0462349358530025), (11, 11.383435258141557), (12, 75.27541204474818), (13, 1.1707886364691775), (14, 5.65384575702712), (15, 1.0736406675386139), (16, 0.22397074028378455), (17, 12.57903351986819), (18, 14.933093640184444), (19, 6.159668420598605), (20, 0.3538989224657254), (21, 0.2430824975484629), (22, 0.29668670784491163), (23, 13.903142409932599), (24, 1.4973346652162456), (25, 3.9983107113681777), (26, 6.25260997211394), (27, 0.009049529535775742), (28, 15.161696617792279), (29, 28.139123499715705), (30, 7.853151014029356), (31, 10.993555804430793), (32, 5.685362944502052), (33, 0.07420538957673493), (34, 4.285267297023855), (35, 18.80721599564378), (36, 4.962638672390888), (37, 0.5545205732810338), (38, 4.709638362052442), (39, 

In [148]:
# set value of threshold to filter features(genes). 
# Here filtered genes with ANOVA f-test scores equal to or higher than 220

counter=0
lst=[]
for i,j in l:
    if j>=220:
        counter+=1 
        lst.append((i,j))
        
        
print(counter) 
print(lst)

# can pick different values for threshold to filter features

### set threshold to 196 to get 11 genes filtered 

5
[(4320, 309.8936581370525), (4435, 264.0930413441192), (11126, 221.71168816092666), (11684, 232.7483560882114), (16255, 230.65716652442927)]


In [149]:
# get column numbers of selected features

tuple_list = lst
first_tuple_elements = []

for a_tuple in tuple_list:
    first_tuple_elements.append(a_tuple[0])
print(first_tuple_elements)

[4320, 4435, 11126, 11684, 16255]


In [150]:
# applying filter to normalized training set
features_train_anova=features_train_scaled.iloc[:, first_tuple_elements]

In [151]:
print(type(features_train_anova))

<class 'pandas.core.frame.DataFrame'>


In [152]:
# applying filter to normalized test set
features_test_anova=features_test_scaled.iloc[:, first_tuple_elements]

In [153]:
print(type(features_train_anova))

<class 'pandas.core.frame.DataFrame'>


In [154]:
# create new train dataset out of final filtered columns(features:genes)


data_train_new= features_train_anova.copy()

print(data_train_new)


NAME   COL10A1    LRRC3B     HHATL     MMP11       CA4
0     0.751747  0.273422  0.202131  0.555992  0.166399
1     0.832066  0.512586  0.259086  0.407746  0.325297
2     0.744690  0.639454  0.172763  0.551670  0.542402
3     0.819112  0.345790  0.407970  0.654505  0.587540
4     0.802032  0.216804  0.160590  0.725568  0.185397
..         ...       ...       ...       ...       ...
361   0.707906  0.333870  0.134754  0.570615  0.207835
362   0.804146  0.270119  0.311400  0.636093  0.681745
363   0.864317  0.129422  0.098351  0.711732  0.205968
364   0.780486  0.708625  0.222099  0.707157  0.339317
365   0.881328  0.309758  0.191709  0.793152  0.338367

[366 rows x 5 columns]


In [155]:

y_train.shape

(366,)

In [156]:
data_train_new['response']= y_train
print(data_train_new)

NAME   COL10A1    LRRC3B     HHATL     MMP11       CA4  response
0     0.751747  0.273422  0.202131  0.555992  0.166399         1
1     0.832066  0.512586  0.259086  0.407746  0.325297         1
2     0.744690  0.639454  0.172763  0.551670  0.542402         1
3     0.819112  0.345790  0.407970  0.654505  0.587540         1
4     0.802032  0.216804  0.160590  0.725568  0.185397         1
..         ...       ...       ...       ...       ...       ...
361   0.707906  0.333870  0.134754  0.570615  0.207835         1
362   0.804146  0.270119  0.311400  0.636093  0.681745         1
363   0.864317  0.129422  0.098351  0.711732  0.205968         1
364   0.780486  0.708625  0.222099  0.707157  0.339317         1
365   0.881328  0.309758  0.191709  0.793152  0.338367         1

[366 rows x 6 columns]


In [157]:
# create new test dataset out of final filtered columns(features:genes)


data_test_new= features_test_anova.copy()
#print(data_train_new.shape)
data_test_new['response']=y_test
print(data_test_new.shape)
print(data_test_new)


(181, 6)
NAME   COL10A1    LRRC3B     HHATL     MMP11       CA4  response
0     0.622232  0.276387  0.228565  0.682729  0.108587         1
1     0.842815  0.291017  0.280945  0.820422  0.296606         1
2     0.711670  0.278669  0.267347  0.729073  0.133592         1
3     0.581285  0.238362  0.115439  0.494055  0.196586         1
4     0.690101  0.289711  0.271146  0.889804  0.205021         1
..         ...       ...       ...       ...       ...       ...
176   0.840046  0.241152  0.241364  0.865220  0.232616         1
177   0.747901  0.363090  0.199361  0.718906  0.153915         1
178   0.866482  0.260210  0.193856  0.862842  0.086571         1
179   0.778352  0.192479  0.210261  0.566795  0.602975         1
180   0.247597  1.000000  0.585395  0.088704  0.952414         0

[181 rows x 6 columns]


In [158]:
X_test_new = data_test_new.iloc[:, :-1]
y_test_new = data_test_new.iloc[:,-1]

X_train_new = data_train_new.iloc[:, :-1]
y_train_new = data_train_new.iloc[:,-1]




In [159]:
X_test_new = X_test_new.to_numpy()
y_test_new = y_test_new.to_numpy()

X_train_new = X_train_new.to_numpy()
y_train_new = y_train_new.to_numpy()



# Fitting models : Random forest, Knn and MLP with GridSearch CV with random over and under sampling . 
## Chosen 0.1 ratio for random over sampling and 0.5 ratio for under sampling so total minority should be in 53 and majority label should be 105 in number(after resampling) from 22(minority) and 525(majority) labels. So now there is 1:2 ratio which is better then original imbalance.

## Can refit models on f1 or roc_auc by changing refit parameter under GridSearchCV and running cells again .

In [160]:
# sklearn RF +random over and undersampling 

from numpy import mean
from sklearn.datasets import make_classification
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.ensemble import RandomForestClassifier

from imblearn.pipeline import Pipeline
from imblearn.over_sampling import RandomOverSampler
from imblearn.under_sampling import RandomUnderSampler
from sklearn.metrics import classification_report
from statistics import *
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import make_scorer, accuracy_score, precision_score, recall_score, f1_score, roc_auc_score

# define pipeline
over = RandomOverSampler(sampling_strategy=0.1,random_state=1)
under = RandomUnderSampler(sampling_strategy=0.5, random_state=1)

steps = [('o', over), ('u', under), ('m', RandomForestClassifier(random_state=1))]
pipeline = Pipeline(steps=steps)



stratified_kfold = RepeatedStratifiedKFold(n_splits=7,random_state=1)
    

param_grid = {
    'm__max_depth': [80,100],'m__max_features': [2, 3],
    'm__min_samples_split': [2,8,10],'m__n_estimators': [100, 200]
}


scoring = {'accuracy': make_scorer(accuracy_score),
           'precision': make_scorer(precision_score, average = 'micro'),
           'recall': make_scorer(recall_score, average = 'micro'),
           'f1': make_scorer(f1_score, average = 'micro'),
           'roc_auc': make_scorer(roc_auc_score)}



grid_search = GridSearchCV(estimator=pipeline,
                           param_grid=param_grid,
                           scoring=scoring,
                           cv=stratified_kfold,refit='roc_auc'
                           )

grid_search.fit(X_train_new, y_train_new)
cv_score = grid_search.best_score_
test_score = grid_search.score(X_test_new, y_test_new)
print(grid_search.best_params_)
print(f'f1 score: {cv_score}\nTest score: {test_score}')

## change refit='f1' under GridSearchCV parameters to refit model using roc_auc



{'m__max_depth': 80, 'm__max_features': 2, 'm__min_samples_split': 2, 'm__n_estimators': 200}
f1 score: 0.9641960784313724
Test score: 0.9942528735632183


In [161]:
# print best parameter after tuning 
print(grid_search.best_params_) 
grid_predictions = grid_search.predict(X_test_new) 
   
# print classification report 
print(classification_report(y_test_new, grid_predictions)) 

{'m__max_depth': 80, 'm__max_features': 2, 'm__min_samples_split': 2, 'm__n_estimators': 200}
              precision    recall  f1-score   support

           0       0.78      1.00      0.88         7
           1       1.00      0.99      0.99       174

    accuracy                           0.99       181
   macro avg       0.89      0.99      0.93       181
weighted avg       0.99      0.99      0.99       181



In [162]:

#evaluating a knn model with random oversampling and undersampling
#k value sqrt(total samples )or plot k vs loss 


from numpy import mean
from sklearn.datasets import make_classification
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedStratifiedKFold
#from sklearn.ensemble import RandomForestClassifier
#from imblearn.ensemble import BalancedRandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from imblearn.pipeline import Pipeline
from imblearn.over_sampling import RandomOverSampler
from imblearn.under_sampling import RandomUnderSampler
from sklearn.metrics import make_scorer, accuracy_score, precision_score, recall_score, f1_score, roc_auc_score



# define pipeline
over = RandomOverSampler(sampling_strategy=0.1, random_state=1)
under = RandomUnderSampler(sampling_strategy=0.5, random_state=1)
steps = [('o', over), ('u', under), ('m', KNeighborsClassifier())]
pipeline = Pipeline(steps=steps)



stratified_kfold = RepeatedStratifiedKFold(n_splits=7,random_state=1)
    


scoring = {'accuracy': make_scorer(accuracy_score),
           'precision': make_scorer(precision_score, average = 'micro'),
           'recall': make_scorer(recall_score, average = 'micro'),
           'f1': make_scorer(f1_score, average = 'micro'),
           'roc_auc': make_scorer(roc_auc_score)}


param_grid = {
    'm__n_neighbors': [5,9,19],'m__weights': ['uniform','distance'],
    'm__metric': ['minkowski','manhattan','euclidean']
}

grid_search = GridSearchCV(estimator=pipeline,
                           param_grid=param_grid,
                           scoring=scoring,
                           cv=stratified_kfold,
                           refit='roc_auc')

grid_search.fit(X_train_new, y_train_new)
cv_score = grid_search.best_score_
test_score = grid_search.score(X_test_new, y_test_new)
print(grid_search.best_params_)
print(f'f1 score: {cv_score}\nTest score: {test_score}')


## change refit='f1' under GridSearchCV parameters to refit model using roc_auc



{'m__metric': 'manhattan', 'm__n_neighbors': 9, 'm__weights': 'uniform'}
f1 score: 0.9995714285714286
Test score: 1.0


In [163]:
# print best parameter after tuning 
print(grid_search.best_params_) 
grid_predictions = grid_search.predict(X_test_new) 
   
# print classification report 
print(classification_report(y_test_new, grid_predictions))  

{'m__metric': 'manhattan', 'm__n_neighbors': 9, 'm__weights': 'uniform'}
              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       1.00      1.00      1.00       174

    accuracy                           1.00       181
   macro avg       1.00      1.00      1.00       181
weighted avg       1.00      1.00      1.00       181



In [164]:
# over and under sampling in MLP classifier 

from numpy import mean
from sklearn.datasets import make_classification
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedStratifiedKFold
#from sklearn.ensemble import RandomForestClassifier
#from imblearn.ensemble import BalancedRandomForestClassifier
#from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
#from sklearn.tree import DecisionTreeClassifier
from imblearn.pipeline import Pipeline
from imblearn.over_sampling import RandomOverSampler
from imblearn.under_sampling import RandomUnderSampler

# defining MLP MODEL 

over = RandomOverSampler(sampling_strategy=0.1,random_state=1)
under = RandomUnderSampler(sampling_strategy=0.5,random_state=1)
steps = [('o', over), ('u', under), ('m', MLPClassifier(validation_fraction=0.1, random_state=1))]
pipeline = Pipeline(steps=steps)
pipeline.fit(X_train_new, y_train_new)

stratified_kfold = RepeatedStratifiedKFold(n_splits=7,random_state=1)

scoring = {'accuracy': make_scorer(accuracy_score),
           'precision': make_scorer(precision_score, average = 'micro'),
           'recall': make_scorer(recall_score, average = 'micro'),
           'f1': make_scorer(f1_score, average = 'micro'),
           'roc_auc': make_scorer(roc_auc_score)}

    

param_grid = {
    'm__hidden_layer_sizes': [100,200,300],'m__alpha': [0.0001,0.001],
    'm__learning_rate': ['constant', 'adaptive'],'m__max_iter':[400,500]
}

grid_search = GridSearchCV(estimator=pipeline,
                           param_grid=param_grid,
                           scoring=scoring,
                           cv=stratified_kfold,refit='roc_auc'
                           )

grid_search.fit(X_train_new, y_train_new)
cv_score = grid_search.best_score_
test_score = grid_search.score(X_test_new, y_test_new)
print(grid_search.best_params_)
print(f'f1 score: {cv_score}\nTest score: {test_score}')



## change refit='roc_auc' under GridSearchCV parameters to refit model using roc_auc



/Users/priyamvadakumar/opt/anaconda3/lib/python3.7/site-packages/sklearn/neural_network/_multilayer_perceptron.py:696: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  ConvergenceWarning,


{'m__alpha': 0.0001, 'm__hidden_layer_sizes': 300, 'm__learning_rate': 'constant', 'm__max_iter': 400}
f1 score: 0.9951456582633053
Test score: 0.9971264367816092


In [165]:
# print best parameter after tuning 
print(grid_search.best_params_) 
grid_predictions = grid_search.predict(X_test_new) 
   
# print classification report 
print(classification_report(y_test_new, grid_predictions)) 

{'m__alpha': 0.0001, 'm__hidden_layer_sizes': 300, 'm__learning_rate': 'constant', 'm__max_iter': 400}
              precision    recall  f1-score   support

           0       0.88      1.00      0.93         7
           1       1.00      0.99      1.00       174

    accuracy                           0.99       181
   macro avg       0.94      1.00      0.97       181
weighted avg       1.00      0.99      0.99       181



# Test set results are actually same whether models are refit with roc_auc or f1 score for the same number of features(genes)

# checking the effect of the filtered genes on response column of test set



In [168]:
#print(data_test_new)

# effect of genes on normal samples 
s = data_test_new[data_test_new.response == 0]
print(s)
# effect of genes on tumor samples 
t = data_test_new[data_test_new.response == 1]
print(t)



NAME   COL10A1    LRRC3B     HHATL     MMP11       CA4  response
25    0.000000  0.866450  0.407173  0.089447  0.844823         0
44    0.036444  0.746623  0.668107  0.090547  1.000000         0
63    0.148485  0.867562  0.570037  0.198692  0.754433         0
86    0.189432  0.833355  0.375107  0.000000  0.987248         0
111   0.128498  0.838547  0.586799  0.025386  0.946238         0
170   0.194421  0.803369  0.538548  0.110107  0.614963         0
180   0.247597  1.000000  0.585395  0.088704  0.952414         0
NAME   COL10A1    LRRC3B     HHATL     MMP11       CA4  response
0     0.622232  0.276387  0.228565  0.682729  0.108587         1
1     0.842815  0.291017  0.280945  0.820422  0.296606         1
2     0.711670  0.278669  0.267347  0.729073  0.133592         1
3     0.581285  0.238362  0.115439  0.494055  0.196586         1
4     0.690101  0.289711  0.271146  0.889804  0.205021         1
..         ...       ...       ...       ...       ...       ...
175   0.756784  0.146437 